Multiple accidentals

erabinov opened the pull request:
<!--
Thanks for contributing a pull request! Please ensure you have taken a look at
the contribution guidelines: https://github.com/librosa/librosa/blob/main/CONTRIBUTING.md#how-to-contribute
-->
#### Reference Issue
Multiple accidentals in tonics, expanded support for modes in note spelling #1724


#### What does this implement/fix? Explain your changes.

I made a first pass at support for multiple accidentals in  `key_to_degrees` and `key_to_notes`.

The `key_to_degrees` method required very little modification.

The basic idea for the approach for `key_to_notes` was: for 'Cbb:min', add flats to each note `key_to_notes('Cb:min')` and then rearrange the scale so that the enharmonic equivalent of C is first.


#### Any other comments?

Some decisions still need to be made regarding whether more than two accidentals are allowed, and the appropriate regular expressions need to be modified.
I have modified one of the existing regular expressions, and added one of my own.

I've also added a few helper methods, namely `note_to_degree` and `simplify_note`.

I've made some comments about places where the code could be made a bit more efficient.

bmcfee commented:
Thanks again for this!  Just leaving a note for now that I'm a bit swamped with life outside of github right now, but hope to find time to properly dig into this soon.  (Don't want to leave you hanging.)

erabinov commented:
Thanks for keeping me posted!

bmcfee commented:
:wave: hi @erabinov , I'm looking into this now, sorry for the delay.

I've taken the liberty of fixing the merge conflict in the AUTHORS.md file so that our CI action can run, but I've noticed that you haven't added or modified any tests yet in this PR.  I tried it out locally, and it seems like your regex doesn't quite work as expected:

```python
In [8]: librosa.key_to_notes(key='B𝄫:maj')
---------------------------------------------------------------------------
IndexError                                Traceback (most recent call last)
Cell In[8], line 1
----> 1 librosa.key_to_notes(key='B𝄫:maj')

File <decorator-gen-4>:2, in key_to_notes(key, unicode)

File ~/miniconda3/envs/py311/lib/python3.11/site-packages/joblib/memory.py:655, in MemorizedFunc.__call__(self, *args, **kwargs)
    654 def __call__(self, *args, **kwargs):
--> 655     return self._cached_call(args, kwargs)[0]

File ~/miniconda3/envs/py311/lib/python3.11/site-packages/joblib/memory.py:598, in MemorizedFunc._cached_call(self, args, kwargs, shelving)
    595     must_call = True
    597 if must_call:
--> 598     out, metadata = self.call(*args, **kwargs)
    599     if self.mmap_mode is not None:
    600         # Memmap the output at the first call to be consistent with
    601         # later calls
    602         if self._verbose:

File ~/miniconda3/envs/py311/lib/python3.11/site-packages/joblib/memory.py:856, in MemorizedFunc.call(self, *args, **kwargs)
    854 if self._verbose > 0:
    855     print(format_call(self.func, args, kwargs))
--> 856 output = self.func(*args, **kwargs)
    857 self.store_backend.dump_item(
    858     [func_id, args_id], output, verbose=self._verbose)
    860 duration = time.time() - start_time

File ~/git/librosa/librosa/core/notation.py:631, in key_to_notes(key, unicode)
    629 notes = deque([simplify_note(note+additional_acc) for note in intermediate_notes])
    630 #Cycle until the equivalent of 'C' is in the first position. This may be a bit inefficient; if additional_acc == +1, then we need to cycle all the way through.
--> 631 while(note_to_degree(notes[0])!=0):
    632     notes.appendleft(notes.pop())
    634 notes = list(notes)

File ~/git/librosa/librosa/core/notation.py:495, in note_to_degree(key)
    492 if not match:
    493     raise ParameterError(f"Improper key format: {key:s}")
--> 495 letter = match.group('tonic').upper()
    496 accidental = match.group('accidental')
    497 pitch_map = {"C": 0, "D": 2, "E": 4, "F": 5, "G": 7, "A": 9, "B": 11}

IndexError: no such group
```

So before I dig into a proper code review here, would it be possible to add a couple of test cases that verify the new functionality is behaving properly?

I'm also curious why a new regex (`UNICODE_NOTE_RE`) was needed for the note parser.  It already supported both unicode and ascii spellings, and it seems like the logic is already there to accumulate multiple accidentals correctly.  The comment you added there indicates that there might be a problem with the previous regexes, but I'd like to know more about what the problem is/was so we can fix it rather than proliferate new parsers.

erabinov commented:
Hi @bmcfee,

Thanks for the feedback! I'll get working on it tomorrow. 

With the regexes, there are two different changes I made.

1. I changed the `KEY_RE` regex to include support for up to two sharps or two flats in the key signature name. If you want more accidentals (if, e.g., you want to continue down the rabbit hole and have support for modes with multiple accidentals), then this would have to change. (Also, this regex as I have it allows `C#b:maj`, which we probably want to exclude.) 
**So it would be helpful to know for sure whether you have in mind to allow, as input to `key_to_notes` function, more than two sharps or flats, and/or sharps and flats in the same key name** (e.g. `C###:min` or `C#b:maj`)

2. The `UNICODE_NOTE_RE`  regex is almost certainly not necessary and I just used it in the helper functions to make my other additions to the code shorter. I'll get rid of it when I work on the code tomorrow.

As for the tests, I added some to the docstring, but I didn't realize I should add them to `tests.py`. I'll work on that tomorrow too.

bmcfee commented on librosa/core/notation.py:
```diff
@@ -122,16 +123,17 @@
 
 
 # Pre-compiled regular expressions for note and key parsing
+KEY_RE = re.compile(
+    r"^(?P<tonic>[A-Ga-g])" r"(?P<accidental>([#♯b!♭𝄪𝄫]*))" r":(?P<scale>(maj|min)(or)?)$"
+)
```
At this point, the first parts of KEY_RE and NOTE_RE are (rightfully) nearly identical, except that KEY doesn't support naturals.  Would it make sense to include that here too?

erabinov replied:
I think you're right, since we have ventured into the forbidden world of key signatures like `C#b##:min`; might as well also allow natural symbols.
```suggestion
KEY_RE = re.compile(
    r"^(?P<tonic>[A-Ga-g])" r"(?P<accidental>([#♯b!♭𝄪𝄫♮]*))" r":(?P<scale>(maj|min)(or)?)$"
)
```

bmcfee commented on librosa/core/notation.py:
```diff
@@ -122,16 +123,17 @@
-KEY_RE = re.compile(
-    r"^(?P<tonic>[A-Ga-g])" r"(?P<accidental>[#♯b!♭]?)" r":(?P<scale>(maj|min)(or)?)$"
-)
+ACC_MAP = {"#": 1, "": 0, "b": -1, "!": -1, "♯": 1, "♭": -1, "##": 2, "♯♯": 2, "𝄪": 2, "bb": -2, "♭♭": -2, "𝄫": -2}
```
I wonder - if we're going to extend the accidental accumulator logic to support variable number of tokens, we can probably remove the explicit support for non-unicode doubles `##` and `bb`, right?

Also, natural should probably include here as mapping explicitly to 0 given the discussion elsewhere in this PR.

erabinov replied:
Agreed. 
```suggestion
ACC_MAP = {"#": 1, "♮": 0, "": 0, "b": -1, "!": -1, "♯": 1, "♭": -1, "𝄪": 2, "𝄫": -2}
```

bmcfee commented on librosa/core/notation.py:
```diff
@@ -472,6 +474,61 @@ def list_thaat() -> List[str]:
     return list(THAAT_MAP.keys())
 
 
+def note_to_degree(key: str) -> int:
```
Is this function meant to be user-facing or purely internal?

If the former, we'll need to expand out the docstring following numpydoc conventions, and add the function to `__all__` and `__init__.pyi`.

If the latter, we should rename this to `__note_to_degree` to convey that the function is intended as private.

erabinov replied:
I intended it only as a helper function, but if you think there could be some use to it to the user, I could make the necessary changes.

erabinov replied:
@bmcfee Let me know what you think about this. I'm happy to do either, but I lean towards keeping the functions private.

bmcfee replied:
I think keeping them private makes sense.

erabinov replied:
Ok, sounds good! I pushed that change to the branch.

bmcfee commented on librosa/core/notation.py:
```diff
@@ -472,6 +474,61 @@ def list_thaat() -> List[str]:
+    return (pitch_map[letter]+sum([ACC_MAP[acc] * counter[acc] for acc in ACC_MAP]))%12
+
+
+def simplify_note(key: str, unicode: bool = True) -> str:
```
same comment as above for `note_to_degree`

bmcfee commented on librosa/core/notation.py:
```diff
@@ -551,16 +608,39 @@ def key_to_notes(key: str, *, unicode: bool = True) -> List[str]:
+        notes = deque([simplify_note(note+additional_acc) for note in intermediate_notes])
+        #Cycle until the equivalent of 'C' is in the first position. This may be a bit inefficient; if additional_acc == +1, then we need to cycle all the way through.
+        while(note_to_degree(notes[0])!=0):
+            notes.appendleft(notes.pop())
```
I feel like this part could be simplified a bit.. some ideas:

Since `notes` is already a `deque` object, you could use `notes.rotate` instead of popping and appending left.  If you could additionally locate the index of the zero in `notes` (which, I believe, will always exist), then you could eliminate the while loop entirely by doing `notes.rotate(index)`.

Now, finding the zero in `notes` is tricky the way the data structure currently works, because the `note_to_degree` function only operates on one note at a time so you have to iteratively call it until you hit the right index.  At the same time, I don't think we actually need the deque object here.  If instead, the `simplify_note` and `note_to_degree` functions were vectorized to support ndarrays of strings (as well as plain strings), then you could so something like:

```python
notes = simplify_note(intermediate_notes + additional_acc)  # May need some tweaking...
degrees = note_to_degree(notes)
notes = np.roll(notes, shift=-np.argwhere(degrees == 0)[0])
```
The first line could require some tweaking because broadcast summing strings like that won't work.  However, since you're creating this simplify note function, you could just add `additional_acc` as a supplemental parameter that gets added in on top of whatever else is in `note`.

For an example of this string vectorization business I'm talking about (to support ndarrays of strings) see
https://github.com/librosa/librosa/blob/09e4a622456bbad314e2e6d5c75879f56d728b9b/librosa/core/notation.py#L879

Basically we add a decorator call and a few extra type hints, but the body of the function doesn't change (much).

erabinov replied:
I like the vectorization idea, but I'm having a bit of trouble making it work locally. 
I tried to add this line of code:
```
@vectorize(otypes = [int])
def note_to_degree(key: str) -> int:
```

When I try to call this method, it converts the strings into byte strings, and then the regex matching stops working:

```
TypeError                                 Traceback (most recent call last)
Cell In[4], line 1
----> 1 simplify_note(np.array(['C##','Cb#']))

File [~/Documents/librosa/librosa/core/notation.py:509](https://file+.vscode-resource.vscode-cdn.net/Users/eugenerabinovich/Documents/~/Documents/librosa/librosa/core/notation.py:509), in simplify_note(key, unicode)
    499 def simplify_note(key: str, unicode: bool = True) -> str:
    500     """Take in a note name and simplify by canceling sharp-flat pairs, and doubling accidentals as appropriate.
    501 
    502     >>> librosa.simplify_note('C♭♯')
   (...)
    507 
    508     """
--> 509     match = NOTE_RE.match(key)
    511     if not match:
    512         raise ParameterError(f"Improper key format: {key:s}")

TypeError: cannot use a string pattern on a bytes-like object
```

Is there a way that I can enforce that the array gets processed as an array of `str` objects instead of `bytes`?

bmcfee replied:
Hmm... maybe I had this wrong in my head.  A more relevant example might be `note_to_midi`:
https://github.com/librosa/librosa/blob/09e4a622456bbad314e2e6d5c75879f56d728b9b/librosa/core/convert.py#L734-L751

which doesn't use numpy vectorize, but rather handles the unrolling recursively.

Sorry for the confusion on that, it's been a minute since I implemented this stuff.  We do use vectorization when going the other way though:
https://github.com/librosa/librosa/blob/09e4a622456bbad314e2e6d5c75879f56d728b9b/librosa/core/convert.py#L899-L900

erabinov replied:
Ok, I made a first pass at these changes. It seemed to be working fine locally, but I may have lost some test coverage, so I may have to do another pass at adding tests.

bmcfee commented:
Thanks @erabinov for adding the tests - this is looking good so far!  I'll try to find some more time later on to see if I can patch this together with my (very broken) prototype for handling modes to see if it correctly handles the cases I was having trouble with (eg `Cb:phrygian`).

I've added some comments on the code now, mainly having to do with some implementation details and consistency with the rest of the package.  I'm not yet sure if the vectorization part is/will be necessary, so I'm curious for your thoughts there.

erabinov commented on librosa/core/notation.py:
```diff
@@ -551,16 +608,39 @@ def key_to_notes(key: str, *, unicode: bool = True) -> List[str]:
+    if multiple:
+        sign_map = {+1: "♯", -1: "♭"}
+        additional_acc = sign_map[np.sign(offset)]
+        intermediate_notes = key_to_notes(tonic+additional_acc*(abs(offset)-1)+':'+scale)
```
It should be peeling off one accidental at a time, with the caveat that mixed sharps and flats are canceled out already, e.g. we process `#b#` as `#`.

erabinov commented:
Ok, @bmcfee, I implemented the discussed changes, leaving open the question of whether the helper methods should be private or public. I may need to make some more tests, but at first attempt in a notebook, the methods were working as I expected.

bmcfee commented:
Thanks @erabinov - looks pretty good!  Ignore the CI failure due to codecov, but we do still have some linting problems.

Meanwhile, I'm (offline) revising my mode->key converter to see how it plays with your implementation.  If that all works smoothly, I think we'll be good to merge.  But I'll still give it one more thorough review before then.

bmcfee commented:
Alright!  Hacked together a mode converter and tested it with the new and improved multi-accidental key code.  Looks to me like everything is working properly!

```python
major_dict = {
    'ion': {'C': 'C', 'D': 'D', 'E': 'E', 'F': 'F', 'G': 'G', 'A': 'A', 'B': 'B'},
    'dor': {'C': 'B♭', 'D': 'C', 'E': 'D', 'F': 'E♭', 'G': 'F', 'A': 'G', 'B': 'A'},
    'phr': {'C': 'A♭', 'D': 'B♭', 'E': 'C', 'F': 'D♭', 'G': 'E♭', 'A': 'F', 'B': 'G'},
    'lyd': {'C': 'G', 'D': 'A', 'E': 'B', 'F': 'C', 'G': 'D', 'A': 'E', 'B': 'F♯'},
    'mix': {'C': 'F', 'D': 'G', 'E': 'A', 'F': 'B♭', 'G': 'C', 'A': 'D', 'B': 'E'},
    'aeo': {'C': 'E♭', 'D': 'F', 'E': 'G', 'F': 'A♭', 'G': 'B♭', 'A': 'C', 'B': 'D'},
    'loc': {'C': 'D♭', 'D': 'E♭', 'E': 'F', 'F': 'G♭', 'G': 'A♭', 'A': 'B♭', 'B': 'C'}
}

def mode_to_key(signature: str) -> str:
    # Translate a mode (eg D:dorian) into its equivalent major key
    
    match = re.match(
        r"^(?P<tonic>[A-Ga-g])"
        r"(?P<accidental>[#♯𝄪b!♭𝄫]*)"
        r":((?P<scale>(maj|min)(or)?)|(?P<mode>(ion|dor|phr(yg)?|lyd|mix(olyd)?|aeol?|locr?)(ian)?))$",
        signature,
    )
    
    if not match:
        raise ParameterError("Improper format: {:s}".format(signature))

    if match.group('scale') or not match.group("mode"):
        # We're already fine here
        return signature
        
    # We have a mode, time to translate
    mode = match.group("mode").lower()[:3]

    # Get the relative major
    tonic = major_dict[mode][match.group("tonic").upper()]

    return f"{tonic}{match.group('accidental')}:maj"
```

The following test code will try all combinations of tonics, single accidentals, and modes.  It will fail if the selected diatonic note names are not unique (ie use all seven names).
```python
for k in 'CDEFGAB':
    for acc in ['b', '', '#']:
        root = k + acc
        for mode in ['ion', 'dor', 'phr', 'lyd', 'mix', 'aeol', 'loc']:
            query = f"{root}:{mode}"
            equiv = mode_to_key(query)

            degs = librosa.key_to_degrees(equiv)
            notes = librosa.key_to_notes(equiv)
            
            names = [notes[_] for _ in degs]
            print(f"{query:10s}\t->\t{notes}\t->\t{names}", end='')
            if len(set([n[0] for n in names])) != 7:
                print('\t->\tFAIL')
            else:
                print()
        print('-' * 15)
```
As demonstrated below, all notes are correctly generated now.  (Woo!)

The only remaining issue here is that they are out of order because we've lost track of what mode we came from.  This is because I've implemented everything outside the key_to_notes function - it first converts mode to key, then gets the degrees from the key.  Properly spelling notes for a mode will require us to rotate the original tonic back into the root position.  Probably that's out of scope for this PR, but worth considering.

<details>
<summary> Output </summary>

```
Cb:ion    	->	['C', 'D♭', 'D', 'E♭', 'F♭', 'F', 'G♭', 'G', 'A♭', 'A', 'B♭', 'C♭']	->	['C♭', 'D♭', 'E♭', 'F♭', 'G♭', 'A♭', 'B♭']
Cb:dor    	->	['D𝄫', 'D♭', 'E𝄫', 'E♭', 'F♭', 'G𝄫', 'G♭', 'A𝄫', 'A♭', 'B𝄫', 'B♭', 'C♭']	->	['B𝄫', 'C♭', 'D♭', 'E𝄫', 'F♭', 'G♭', 'A♭']
Cb:phr    	->	['D𝄫', 'D♭', 'E𝄫', 'E♭', 'F♭', 'G𝄫', 'G♭', 'A𝄫', 'A♭', 'B𝄫', 'B♭', 'C♭']	->	['A𝄫', 'B𝄫', 'C♭', 'D𝄫', 'E𝄫', 'F♭', 'G♭']
Cb:lyd    	->	['C', 'D♭', 'D', 'E♭', 'E', 'F', 'G♭', 'G', 'A♭', 'A', 'B♭', 'C♭']	->	['G♭', 'A♭', 'B♭', 'C♭', 'D♭', 'E♭', 'F']
Cb:mix    	->	['C', 'D♭', 'D', 'E♭', 'F♭', 'F', 'G♭', 'G', 'A♭', 'B𝄫', 'B♭', 'C♭']	->	['F♭', 'G♭', 'A♭', 'B𝄫', 'C♭', 'D♭', 'E♭']
Cb:aeol   	->	['D𝄫', 'D♭', 'E𝄫', 'E♭', 'F♭', 'G𝄫', 'G♭', 'A𝄫', 'A♭', 'B𝄫', 'B♭', 'C♭']	->	['E𝄫', 'F♭', 'G♭', 'A𝄫', 'B𝄫', 'C♭', 'D♭']
Cb:loc    	->	['D𝄫', 'D♭', 'E𝄫', 'E♭', 'F♭', 'G𝄫', 'G♭', 'A𝄫', 'A♭', 'B𝄫', 'B♭', 'C♭']	->	['D𝄫', 'E𝄫', 'F♭', 'G𝄫', 'A𝄫', 'B𝄫', 'C♭']
---------------
C:ion     	->	['C', 'C♯', 'D', 'D♯', 'E', 'F', 'F♯', 'G', 'G♯', 'A', 'A♯', 'B']	->	['C', 'D', 'E', 'F', 'G', 'A', 'B']
C:dor     	->	['C', 'D♭', 'D', 'E♭', 'E', 'F', 'G♭', 'G', 'A♭', 'A', 'B♭', 'B']	->	['B♭', 'C', 'D', 'E♭', 'F', 'G', 'A']
C:phr     	->	['C', 'D♭', 'D', 'E♭', 'E', 'F', 'G♭', 'G', 'A♭', 'A', 'B♭', 'B']	->	['A♭', 'B♭', 'C', 'D♭', 'E♭', 'F', 'G']
C:lyd     	->	['C', 'C♯', 'D', 'D♯', 'E', 'F', 'F♯', 'G', 'G♯', 'A', 'A♯', 'B']	->	['G', 'A', 'B', 'C', 'D', 'E', 'F♯']
C:mix     	->	['C', 'D♭', 'D', 'E♭', 'E', 'F', 'G♭', 'G', 'A♭', 'A', 'B♭', 'B']	->	['F', 'G', 'A', 'B♭', 'C', 'D', 'E']
C:aeol    	->	['C', 'D♭', 'D', 'E♭', 'E', 'F', 'G♭', 'G', 'A♭', 'A', 'B♭', 'B']	->	['E♭', 'F', 'G', 'A♭', 'B♭', 'C', 'D']
C:loc     	->	['C', 'D♭', 'D', 'E♭', 'E', 'F', 'G♭', 'G', 'A♭', 'A', 'B♭', 'B']	->	['D♭', 'E♭', 'F', 'G♭', 'A♭', 'B♭', 'C']
---------------
C#:ion    	->	['B♯', 'C♯', 'D', 'D♯', 'E', 'E♯', 'F♯', 'G', 'G♯', 'A', 'A♯', 'B']	->	['C♯', 'D♯', 'E♯', 'F♯', 'G♯', 'A♯', 'B♯']
C#:dor    	->	['C', 'C♯', 'D', 'D♯', 'E', 'F', 'F♯', 'G', 'G♯', 'A', 'A♯', 'B']	->	['B', 'C♯', 'D♯', 'E', 'F♯', 'G♯', 'A♯']
C#:phr    	->	['C', 'C♯', 'D', 'D♯', 'E', 'F', 'F♯', 'G', 'G♯', 'A', 'A♯', 'B']	->	['A', 'B', 'C♯', 'D', 'E', 'F♯', 'G♯']
C#:lyd    	->	['B♯', 'C♯', 'D', 'D♯', 'E', 'E♯', 'F♯', 'F𝄪', 'G♯', 'A', 'A♯', 'B']	->	['G♯', 'A♯', 'B♯', 'C♯', 'D♯', 'E♯', 'F𝄪']
C#:mix    	->	['C', 'C♯', 'D', 'D♯', 'E', 'E♯', 'F♯', 'G', 'G♯', 'A', 'A♯', 'B']	->	['F♯', 'G♯', 'A♯', 'B', 'C♯', 'D♯', 'E♯']
C#:aeol   	->	['C', 'C♯', 'D', 'D♯', 'E', 'F', 'F♯', 'G', 'G♯', 'A', 'A♯', 'B']	->	['E', 'F♯', 'G♯', 'A', 'B', 'C♯', 'D♯']
C#:loc    	->	['C', 'C♯', 'D', 'D♯', 'E', 'F', 'F♯', 'G', 'G♯', 'A', 'A♯', 'B']	->	['D', 'E', 'F♯', 'G', 'A', 'B', 'C♯']
---------------
Db:ion    	->	['C', 'D♭', 'D', 'E♭', 'E', 'F', 'G♭', 'G', 'A♭', 'A', 'B♭', 'B']	->	['D♭', 'E♭', 'F', 'G♭', 'A♭', 'B♭', 'C']
Db:dor    	->	['C', 'D♭', 'D', 'E♭', 'F♭', 'F', 'G♭', 'G', 'A♭', 'A', 'B♭', 'C♭']	->	['C♭', 'D♭', 'E♭', 'F♭', 'G♭', 'A♭', 'B♭']
Db:phr    	->	['D𝄫', 'D♭', 'E𝄫', 'E♭', 'F♭', 'G𝄫', 'G♭', 'A𝄫', 'A♭', 'B𝄫', 'B♭', 'C♭']	->	['B𝄫', 'C♭', 'D♭', 'E𝄫', 'F♭', 'G♭', 'A♭']
Db:lyd    	->	['C', 'D♭', 'D', 'E♭', 'E', 'F', 'G♭', 'G', 'A♭', 'A', 'B♭', 'B']	->	['A♭', 'B♭', 'C', 'D♭', 'E♭', 'F', 'G']
Db:mix    	->	['C', 'D♭', 'D', 'E♭', 'E', 'F', 'G♭', 'G', 'A♭', 'A', 'B♭', 'C♭']	->	['G♭', 'A♭', 'B♭', 'C♭', 'D♭', 'E♭', 'F']
Db:aeol   	->	['C', 'D♭', 'D', 'E♭', 'F♭', 'F', 'G♭', 'G', 'A♭', 'B𝄫', 'B♭', 'C♭']	->	['F♭', 'G♭', 'A♭', 'B𝄫', 'C♭', 'D♭', 'E♭']
Db:loc    	->	['D𝄫', 'D♭', 'E𝄫', 'E♭', 'F♭', 'G𝄫', 'G♭', 'A𝄫', 'A♭', 'B𝄫', 'B♭', 'C♭']	->	['E𝄫', 'F♭', 'G♭', 'A𝄫', 'B𝄫', 'C♭', 'D♭']
---------------
D:ion     	->	['C', 'C♯', 'D', 'D♯', 'E', 'F', 'F♯', 'G', 'G♯', 'A', 'A♯', 'B']	->	['D', 'E', 'F♯', 'G', 'A', 'B', 'C♯']
D:dor     	->	['C', 'C♯', 'D', 'D♯', 'E', 'F', 'F♯', 'G', 'G♯', 'A', 'A♯', 'B']	->	['C', 'D', 'E', 'F', 'G', 'A', 'B']
D:phr     	->	['C', 'D♭', 'D', 'E♭', 'E', 'F', 'G♭', 'G', 'A♭', 'A', 'B♭', 'B']	->	['B♭', 'C', 'D', 'E♭', 'F', 'G', 'A']
D:lyd     	->	['C', 'C♯', 'D', 'D♯', 'E', 'F', 'F♯', 'G', 'G♯', 'A', 'A♯', 'B']	->	['A', 'B', 'C♯', 'D', 'E', 'F♯', 'G♯']
D:mix     	->	['C', 'C♯', 'D', 'D♯', 'E', 'F', 'F♯', 'G', 'G♯', 'A', 'A♯', 'B']	->	['G', 'A', 'B', 'C', 'D', 'E', 'F♯']
D:aeol    	->	['C', 'D♭', 'D', 'E♭', 'E', 'F', 'G♭', 'G', 'A♭', 'A', 'B♭', 'B']	->	['F', 'G', 'A', 'B♭', 'C', 'D', 'E']
D:loc     	->	['C', 'D♭', 'D', 'E♭', 'E', 'F', 'G♭', 'G', 'A♭', 'A', 'B♭', 'B']	->	['E♭', 'F', 'G', 'A♭', 'B♭', 'C', 'D']
---------------
D#:ion    	->	['B♯', 'C♯', 'C𝄪', 'D♯', 'E', 'E♯', 'F♯', 'F𝄪', 'G♯', 'A', 'A♯', 'B']	->	['D♯', 'E♯', 'F𝄪', 'G♯', 'A♯', 'B♯', 'C𝄪']
D#:dor    	->	['B♯', 'C♯', 'D', 'D♯', 'E', 'E♯', 'F♯', 'G', 'G♯', 'A', 'A♯', 'B']	->	['C♯', 'D♯', 'E♯', 'F♯', 'G♯', 'A♯', 'B♯']
D#:phr    	->	['C', 'C♯', 'D', 'D♯', 'E', 'F', 'F♯', 'G', 'G♯', 'A', 'A♯', 'B']	->	['B', 'C♯', 'D♯', 'E', 'F♯', 'G♯', 'A♯']
D#:lyd    	->	['B♯', 'C♯', 'C𝄪', 'D♯', 'E', 'E♯', 'F♯', 'F𝄪', 'G♯', 'G𝄪', 'A♯', 'B']	->	['A♯', 'B♯', 'C𝄪', 'D♯', 'E♯', 'F𝄪', 'G𝄪']
D#:mix    	->	['B♯', 'C♯', 'D', 'D♯', 'E', 'E♯', 'F♯', 'F𝄪', 'G♯', 'A', 'A♯', 'B']	->	['G♯', 'A♯', 'B♯', 'C♯', 'D♯', 'E♯', 'F𝄪']
D#:aeol   	->	['C', 'C♯', 'D', 'D♯', 'E', 'E♯', 'F♯', 'G', 'G♯', 'A', 'A♯', 'B']	->	['F♯', 'G♯', 'A♯', 'B', 'C♯', 'D♯', 'E♯']
D#:loc    	->	['C', 'C♯', 'D', 'D♯', 'E', 'F', 'F♯', 'G', 'G♯', 'A', 'A♯', 'B']	->	['E', 'F♯', 'G♯', 'A', 'B', 'C♯', 'D♯']
---------------
Eb:ion    	->	['C', 'D♭', 'D', 'E♭', 'E', 'F', 'G♭', 'G', 'A♭', 'A', 'B♭', 'B']	->	['E♭', 'F', 'G', 'A♭', 'B♭', 'C', 'D']
Eb:dor    	->	['C', 'D♭', 'D', 'E♭', 'E', 'F', 'G♭', 'G', 'A♭', 'A', 'B♭', 'B']	->	['D♭', 'E♭', 'F', 'G♭', 'A♭', 'B♭', 'C']
Eb:phr    	->	['C', 'D♭', 'D', 'E♭', 'F♭', 'F', 'G♭', 'G', 'A♭', 'A', 'B♭', 'C♭']	->	['C♭', 'D♭', 'E♭', 'F♭', 'G♭', 'A♭', 'B♭']
Eb:lyd    	->	['C', 'D♭', 'D', 'E♭', 'E', 'F', 'G♭', 'G', 'A♭', 'A', 'B♭', 'B']	->	['B♭', 'C', 'D', 'E♭', 'F', 'G', 'A']
Eb:mix    	->	['C', 'D♭', 'D', 'E♭', 'E', 'F', 'G♭', 'G', 'A♭', 'A', 'B♭', 'B']	->	['A♭', 'B♭', 'C', 'D♭', 'E♭', 'F', 'G']
Eb:aeol   	->	['C', 'D♭', 'D', 'E♭', 'E', 'F', 'G♭', 'G', 'A♭', 'A', 'B♭', 'C♭']	->	['G♭', 'A♭', 'B♭', 'C♭', 'D♭', 'E♭', 'F']
Eb:loc    	->	['C', 'D♭', 'D', 'E♭', 'F♭', 'F', 'G♭', 'G', 'A♭', 'B𝄫', 'B♭', 'C♭']	->	['F♭', 'G♭', 'A♭', 'B𝄫', 'C♭', 'D♭', 'E♭']
---------------
E:ion     	->	['C', 'C♯', 'D', 'D♯', 'E', 'F', 'F♯', 'G', 'G♯', 'A', 'A♯', 'B']	->	['E', 'F♯', 'G♯', 'A', 'B', 'C♯', 'D♯']
E:dor     	->	['C', 'C♯', 'D', 'D♯', 'E', 'F', 'F♯', 'G', 'G♯', 'A', 'A♯', 'B']	->	['D', 'E', 'F♯', 'G', 'A', 'B', 'C♯']
E:phr     	->	['C', 'C♯', 'D', 'D♯', 'E', 'F', 'F♯', 'G', 'G♯', 'A', 'A♯', 'B']	->	['C', 'D', 'E', 'F', 'G', 'A', 'B']
E:lyd     	->	['C', 'C♯', 'D', 'D♯', 'E', 'F', 'F♯', 'G', 'G♯', 'A', 'A♯', 'B']	->	['B', 'C♯', 'D♯', 'E', 'F♯', 'G♯', 'A♯']
E:mix     	->	['C', 'C♯', 'D', 'D♯', 'E', 'F', 'F♯', 'G', 'G♯', 'A', 'A♯', 'B']	->	['A', 'B', 'C♯', 'D', 'E', 'F♯', 'G♯']
E:aeol    	->	['C', 'C♯', 'D', 'D♯', 'E', 'F', 'F♯', 'G', 'G♯', 'A', 'A♯', 'B']	->	['G', 'A', 'B', 'C', 'D', 'E', 'F♯']
E:loc     	->	['C', 'D♭', 'D', 'E♭', 'E', 'F', 'G♭', 'G', 'A♭', 'A', 'B♭', 'B']	->	['F', 'G', 'A', 'B♭', 'C', 'D', 'E']
---------------
E#:ion    	->	['B♯', 'C♯', 'C𝄪', 'D♯', 'D𝄪', 'E♯', 'F♯', 'F𝄪', 'G♯', 'G𝄪', 'A♯', 'B']	->	['E♯', 'F𝄪', 'G𝄪', 'A♯', 'B♯', 'C𝄪', 'D𝄪']
E#:dor    	->	['B♯', 'C♯', 'C𝄪', 'D♯', 'E', 'E♯', 'F♯', 'F𝄪', 'G♯', 'A', 'A♯', 'B']	->	['D♯', 'E♯', 'F𝄪', 'G♯', 'A♯', 'B♯', 'C𝄪']
E#:phr    	->	['B♯', 'C♯', 'D', 'D♯', 'E', 'E♯', 'F♯', 'G', 'G♯', 'A', 'A♯', 'B']	->	['C♯', 'D♯', 'E♯', 'F♯', 'G♯', 'A♯', 'B♯']
E#:lyd    	->	['B♯', 'C♯', 'C𝄪', 'D♯', 'D𝄪', 'E♯', 'F♯', 'F𝄪', 'G♯', 'G𝄪', 'A♯', 'A𝄪']	->	['B♯', 'C𝄪', 'D𝄪', 'E♯', 'F𝄪', 'G𝄪', 'A𝄪']
E#:mix    	->	['B♯', 'C♯', 'C𝄪', 'D♯', 'E', 'E♯', 'F♯', 'F𝄪', 'G♯', 'G𝄪', 'A♯', 'B']	->	['A♯', 'B♯', 'C𝄪', 'D♯', 'E♯', 'F𝄪', 'G𝄪']
E#:aeol   	->	['B♯', 'C♯', 'D', 'D♯', 'E', 'E♯', 'F♯', 'F𝄪', 'G♯', 'A', 'A♯', 'B']	->	['G♯', 'A♯', 'B♯', 'C♯', 'D♯', 'E♯', 'F𝄪']
E#:loc    	->	['C', 'C♯', 'D', 'D♯', 'E', 'E♯', 'F♯', 'G', 'G♯', 'A', 'A♯', 'B']	->	['F♯', 'G♯', 'A♯', 'B', 'C♯', 'D♯', 'E♯']
---------------
Fb:ion    	->	['C', 'D♭', 'D', 'E♭', 'F♭', 'F', 'G♭', 'G', 'A♭', 'B𝄫', 'B♭', 'C♭']	->	['F♭', 'G♭', 'A♭', 'B𝄫', 'C♭', 'D♭', 'E♭']
Fb:dor    	->	['D𝄫', 'D♭', 'E𝄫', 'E♭', 'F♭', 'G𝄫', 'G♭', 'A𝄫', 'A♭', 'B𝄫', 'B♭', 'C♭']	->	['E𝄫', 'F♭', 'G♭', 'A𝄫', 'B𝄫', 'C♭', 'D♭']
Fb:phr    	->	['D𝄫', 'D♭', 'E𝄫', 'E♭', 'F♭', 'G𝄫', 'G♭', 'A𝄫', 'A♭', 'B𝄫', 'B♭', 'C♭']	->	['D𝄫', 'E𝄫', 'F♭', 'G𝄫', 'A𝄫', 'B𝄫', 'C♭']
Fb:lyd    	->	['C', 'D♭', 'D', 'E♭', 'F♭', 'F', 'G♭', 'G', 'A♭', 'A', 'B♭', 'C♭']	->	['C♭', 'D♭', 'E♭', 'F♭', 'G♭', 'A♭', 'B♭']
Fb:mix    	->	['D𝄫', 'D♭', 'E𝄫', 'E♭', 'F♭', 'G𝄫', 'G♭', 'A𝄫', 'A♭', 'B𝄫', 'B♭', 'C♭']	->	['B𝄫', 'C♭', 'D♭', 'E𝄫', 'F♭', 'G♭', 'A♭']
Fb:aeol   	->	['D𝄫', 'D♭', 'E𝄫', 'E♭', 'F♭', 'G𝄫', 'G♭', 'A𝄫', 'A♭', 'B𝄫', 'B♭', 'C♭']	->	['A𝄫', 'B𝄫', 'C♭', 'D𝄫', 'E𝄫', 'F♭', 'G♭']
Fb:loc    	->	['D𝄫', 'D♭', 'E𝄫', 'E♭', 'F♭', 'G𝄫', 'G♭', 'A𝄫', 'A♭', 'B𝄫', 'C𝄫', 'C♭']	->	['G𝄫', 'A𝄫', 'B𝄫', 'C𝄫', 'D𝄫', 'E𝄫', 'F♭']
---------------
F:ion     	->	['C', 'D♭', 'D', 'E♭', 'E', 'F', 'G♭', 'G', 'A♭', 'A', 'B♭', 'B']	->	['F', 'G', 'A', 'B♭', 'C', 'D', 'E']
F:dor     	->	['C', 'D♭', 'D', 'E♭', 'E', 'F', 'G♭', 'G', 'A♭', 'A', 'B♭', 'B']	->	['E♭', 'F', 'G', 'A♭', 'B♭', 'C', 'D']
F:phr     	->	['C', 'D♭', 'D', 'E♭', 'E', 'F', 'G♭', 'G', 'A♭', 'A', 'B♭', 'B']	->	['D♭', 'E♭', 'F', 'G♭', 'A♭', 'B♭', 'C']
F:lyd     	->	['C', 'C♯', 'D', 'D♯', 'E', 'F', 'F♯', 'G', 'G♯', 'A', 'A♯', 'B']	->	['C', 'D', 'E', 'F', 'G', 'A', 'B']
F:mix     	->	['C', 'D♭', 'D', 'E♭', 'E', 'F', 'G♭', 'G', 'A♭', 'A', 'B♭', 'B']	->	['B♭', 'C', 'D', 'E♭', 'F', 'G', 'A']
F:aeol    	->	['C', 'D♭', 'D', 'E♭', 'E', 'F', 'G♭', 'G', 'A♭', 'A', 'B♭', 'B']	->	['A♭', 'B♭', 'C', 'D♭', 'E♭', 'F', 'G']
F:loc     	->	['C', 'D♭', 'D', 'E♭', 'E', 'F', 'G♭', 'G', 'A♭', 'A', 'B♭', 'C♭']	->	['G♭', 'A♭', 'B♭', 'C♭', 'D♭', 'E♭', 'F']
---------------
F#:ion    	->	['C', 'C♯', 'D', 'D♯', 'E', 'E♯', 'F♯', 'G', 'G♯', 'A', 'A♯', 'B']	->	['F♯', 'G♯', 'A♯', 'B', 'C♯', 'D♯', 'E♯']
F#:dor    	->	['C', 'C♯', 'D', 'D♯', 'E', 'F', 'F♯', 'G', 'G♯', 'A', 'A♯', 'B']	->	['E', 'F♯', 'G♯', 'A', 'B', 'C♯', 'D♯']
F#:phr    	->	['C', 'C♯', 'D', 'D♯', 'E', 'F', 'F♯', 'G', 'G♯', 'A', 'A♯', 'B']	->	['D', 'E', 'F♯', 'G', 'A', 'B', 'C♯']
F#:lyd    	->	['B♯', 'C♯', 'D', 'D♯', 'E', 'E♯', 'F♯', 'G', 'G♯', 'A', 'A♯', 'B']	->	['C♯', 'D♯', 'E♯', 'F♯', 'G♯', 'A♯', 'B♯']
F#:mix    	->	['C', 'C♯', 'D', 'D♯', 'E', 'F', 'F♯', 'G', 'G♯', 'A', 'A♯', 'B']	->	['B', 'C♯', 'D♯', 'E', 'F♯', 'G♯', 'A♯']
F#:aeol   	->	['C', 'C♯', 'D', 'D♯', 'E', 'F', 'F♯', 'G', 'G♯', 'A', 'A♯', 'B']	->	['A', 'B', 'C♯', 'D', 'E', 'F♯', 'G♯']
F#:loc    	->	['C', 'C♯', 'D', 'D♯', 'E', 'F', 'F♯', 'G', 'G♯', 'A', 'A♯', 'B']	->	['G', 'A', 'B', 'C', 'D', 'E', 'F♯']
---------------
Gb:ion    	->	['C', 'D♭', 'D', 'E♭', 'E', 'F', 'G♭', 'G', 'A♭', 'A', 'B♭', 'C♭']	->	['G♭', 'A♭', 'B♭', 'C♭', 'D♭', 'E♭', 'F']
Gb:dor    	->	['C', 'D♭', 'D', 'E♭', 'F♭', 'F', 'G♭', 'G', 'A♭', 'B𝄫', 'B♭', 'C♭']	->	['F♭', 'G♭', 'A♭', 'B𝄫', 'C♭', 'D♭', 'E♭']
Gb:phr    	->	['D𝄫', 'D♭', 'E𝄫', 'E♭', 'F♭', 'G𝄫', 'G♭', 'A𝄫', 'A♭', 'B𝄫', 'B♭', 'C♭']	->	['E𝄫', 'F♭', 'G♭', 'A𝄫', 'B𝄫', 'C♭', 'D♭']
Gb:lyd    	->	['C', 'D♭', 'D', 'E♭', 'E', 'F', 'G♭', 'G', 'A♭', 'A', 'B♭', 'B']	->	['D♭', 'E♭', 'F', 'G♭', 'A♭', 'B♭', 'C']
Gb:mix    	->	['C', 'D♭', 'D', 'E♭', 'F♭', 'F', 'G♭', 'G', 'A♭', 'A', 'B♭', 'C♭']	->	['C♭', 'D♭', 'E♭', 'F♭', 'G♭', 'A♭', 'B♭']
Gb:aeol   	->	['D𝄫', 'D♭', 'E𝄫', 'E♭', 'F♭', 'G𝄫', 'G♭', 'A𝄫', 'A♭', 'B𝄫', 'B♭', 'C♭']	->	['B𝄫', 'C♭', 'D♭', 'E𝄫', 'F♭', 'G♭', 'A♭']
Gb:loc    	->	['D𝄫', 'D♭', 'E𝄫', 'E♭', 'F♭', 'G𝄫', 'G♭', 'A𝄫', 'A♭', 'B𝄫', 'B♭', 'C♭']	->	['A𝄫', 'B𝄫', 'C♭', 'D𝄫', 'E𝄫', 'F♭', 'G♭']
---------------
G:ion     	->	['C', 'C♯', 'D', 'D♯', 'E', 'F', 'F♯', 'G', 'G♯', 'A', 'A♯', 'B']	->	['G', 'A', 'B', 'C', 'D', 'E', 'F♯']
G:dor     	->	['C', 'D♭', 'D', 'E♭', 'E', 'F', 'G♭', 'G', 'A♭', 'A', 'B♭', 'B']	->	['F', 'G', 'A', 'B♭', 'C', 'D', 'E']
G:phr     	->	['C', 'D♭', 'D', 'E♭', 'E', 'F', 'G♭', 'G', 'A♭', 'A', 'B♭', 'B']	->	['E♭', 'F', 'G', 'A♭', 'B♭', 'C', 'D']
G:lyd     	->	['C', 'C♯', 'D', 'D♯', 'E', 'F', 'F♯', 'G', 'G♯', 'A', 'A♯', 'B']	->	['D', 'E', 'F♯', 'G', 'A', 'B', 'C♯']
G:mix     	->	['C', 'C♯', 'D', 'D♯', 'E', 'F', 'F♯', 'G', 'G♯', 'A', 'A♯', 'B']	->	['C', 'D', 'E', 'F', 'G', 'A', 'B']
G:aeol    	->	['C', 'D♭', 'D', 'E♭', 'E', 'F', 'G♭', 'G', 'A♭', 'A', 'B♭', 'B']	->	['B♭', 'C', 'D', 'E♭', 'F', 'G', 'A']
G:loc     	->	['C', 'D♭', 'D', 'E♭', 'E', 'F', 'G♭', 'G', 'A♭', 'A', 'B♭', 'B']	->	['A♭', 'B♭', 'C', 'D♭', 'E♭', 'F', 'G']
---------------
G#:ion    	->	['B♯', 'C♯', 'D', 'D♯', 'E', 'E♯', 'F♯', 'F𝄪', 'G♯', 'A', 'A♯', 'B']	->	['G♯', 'A♯', 'B♯', 'C♯', 'D♯', 'E♯', 'F𝄪']
G#:dor    	->	['C', 'C♯', 'D', 'D♯', 'E', 'E♯', 'F♯', 'G', 'G♯', 'A', 'A♯', 'B']	->	['F♯', 'G♯', 'A♯', 'B', 'C♯', 'D♯', 'E♯']
G#:phr    	->	['C', 'C♯', 'D', 'D♯', 'E', 'F', 'F♯', 'G', 'G♯', 'A', 'A♯', 'B']	->	['E', 'F♯', 'G♯', 'A', 'B', 'C♯', 'D♯']
G#:lyd    	->	['B♯', 'C♯', 'C𝄪', 'D♯', 'E', 'E♯', 'F♯', 'F𝄪', 'G♯', 'A', 'A♯', 'B']	->	['D♯', 'E♯', 'F𝄪', 'G♯', 'A♯', 'B♯', 'C𝄪']
G#:mix    	->	['B♯', 'C♯', 'D', 'D♯', 'E', 'E♯', 'F♯', 'G', 'G♯', 'A', 'A♯', 'B']	->	['C♯', 'D♯', 'E♯', 'F♯', 'G♯', 'A♯', 'B♯']
G#:aeol   	->	['C', 'C♯', 'D', 'D♯', 'E', 'F', 'F♯', 'G', 'G♯', 'A', 'A♯', 'B']	->	['B', 'C♯', 'D♯', 'E', 'F♯', 'G♯', 'A♯']
G#:loc    	->	['C', 'C♯', 'D', 'D♯', 'E', 'F', 'F♯', 'G', 'G♯', 'A', 'A♯', 'B']	->	['A', 'B', 'C♯', 'D', 'E', 'F♯', 'G♯']
---------------
Ab:ion    	->	['C', 'D♭', 'D', 'E♭', 'E', 'F', 'G♭', 'G', 'A♭', 'A', 'B♭', 'B']	->	['A♭', 'B♭', 'C', 'D♭', 'E♭', 'F', 'G']
Ab:dor    	->	['C', 'D♭', 'D', 'E♭', 'E', 'F', 'G♭', 'G', 'A♭', 'A', 'B♭', 'C♭']	->	['G♭', 'A♭', 'B♭', 'C♭', 'D♭', 'E♭', 'F']
Ab:phr    	->	['C', 'D♭', 'D', 'E♭', 'F♭', 'F', 'G♭', 'G', 'A♭', 'B𝄫', 'B♭', 'C♭']	->	['F♭', 'G♭', 'A♭', 'B𝄫', 'C♭', 'D♭', 'E♭']
Ab:lyd    	->	['C', 'D♭', 'D', 'E♭', 'E', 'F', 'G♭', 'G', 'A♭', 'A', 'B♭', 'B']	->	['E♭', 'F', 'G', 'A♭', 'B♭', 'C', 'D']
Ab:mix    	->	['C', 'D♭', 'D', 'E♭', 'E', 'F', 'G♭', 'G', 'A♭', 'A', 'B♭', 'B']	->	['D♭', 'E♭', 'F', 'G♭', 'A♭', 'B♭', 'C']
Ab:aeol   	->	['C', 'D♭', 'D', 'E♭', 'F♭', 'F', 'G♭', 'G', 'A♭', 'A', 'B♭', 'C♭']	->	['C♭', 'D♭', 'E♭', 'F♭', 'G♭', 'A♭', 'B♭']
Ab:loc    	->	['D𝄫', 'D♭', 'E𝄫', 'E♭', 'F♭', 'G𝄫', 'G♭', 'A𝄫', 'A♭', 'B𝄫', 'B♭', 'C♭']	->	['B𝄫', 'C♭', 'D♭', 'E𝄫', 'F♭', 'G♭', 'A♭']
---------------
A:ion     	->	['C', 'C♯', 'D', 'D♯', 'E', 'F', 'F♯', 'G', 'G♯', 'A', 'A♯', 'B']	->	['A', 'B', 'C♯', 'D', 'E', 'F♯', 'G♯']
A:dor     	->	['C', 'C♯', 'D', 'D♯', 'E', 'F', 'F♯', 'G', 'G♯', 'A', 'A♯', 'B']	->	['G', 'A', 'B', 'C', 'D', 'E', 'F♯']
A:phr     	->	['C', 'D♭', 'D', 'E♭', 'E', 'F', 'G♭', 'G', 'A♭', 'A', 'B♭', 'B']	->	['F', 'G', 'A', 'B♭', 'C', 'D', 'E']
A:lyd     	->	['C', 'C♯', 'D', 'D♯', 'E', 'F', 'F♯', 'G', 'G♯', 'A', 'A♯', 'B']	->	['E', 'F♯', 'G♯', 'A', 'B', 'C♯', 'D♯']
A:mix     	->	['C', 'C♯', 'D', 'D♯', 'E', 'F', 'F♯', 'G', 'G♯', 'A', 'A♯', 'B']	->	['D', 'E', 'F♯', 'G', 'A', 'B', 'C♯']
A:aeol    	->	['C', 'C♯', 'D', 'D♯', 'E', 'F', 'F♯', 'G', 'G♯', 'A', 'A♯', 'B']	->	['C', 'D', 'E', 'F', 'G', 'A', 'B']
A:loc     	->	['C', 'D♭', 'D', 'E♭', 'E', 'F', 'G♭', 'G', 'A♭', 'A', 'B♭', 'B']	->	['B♭', 'C', 'D', 'E♭', 'F', 'G', 'A']
---------------
A#:ion    	->	['B♯', 'C♯', 'C𝄪', 'D♯', 'E', 'E♯', 'F♯', 'F𝄪', 'G♯', 'G𝄪', 'A♯', 'B']	->	['A♯', 'B♯', 'C𝄪', 'D♯', 'E♯', 'F𝄪', 'G𝄪']
A#:dor    	->	['B♯', 'C♯', 'D', 'D♯', 'E', 'E♯', 'F♯', 'F𝄪', 'G♯', 'A', 'A♯', 'B']	->	['G♯', 'A♯', 'B♯', 'C♯', 'D♯', 'E♯', 'F𝄪']
A#:phr    	->	['C', 'C♯', 'D', 'D♯', 'E', 'E♯', 'F♯', 'G', 'G♯', 'A', 'A♯', 'B']	->	['F♯', 'G♯', 'A♯', 'B', 'C♯', 'D♯', 'E♯']
A#:lyd    	->	['B♯', 'C♯', 'C𝄪', 'D♯', 'D𝄪', 'E♯', 'F♯', 'F𝄪', 'G♯', 'G𝄪', 'A♯', 'B']	->	['E♯', 'F𝄪', 'G𝄪', 'A♯', 'B♯', 'C𝄪', 'D𝄪']
A#:mix    	->	['B♯', 'C♯', 'C𝄪', 'D♯', 'E', 'E♯', 'F♯', 'F𝄪', 'G♯', 'A', 'A♯', 'B']	->	['D♯', 'E♯', 'F𝄪', 'G♯', 'A♯', 'B♯', 'C𝄪']
A#:aeol   	->	['B♯', 'C♯', 'D', 'D♯', 'E', 'E♯', 'F♯', 'G', 'G♯', 'A', 'A♯', 'B']	->	['C♯', 'D♯', 'E♯', 'F♯', 'G♯', 'A♯', 'B♯']
A#:loc    	->	['C', 'C♯', 'D', 'D♯', 'E', 'F', 'F♯', 'G', 'G♯', 'A', 'A♯', 'B']	->	['B', 'C♯', 'D♯', 'E', 'F♯', 'G♯', 'A♯']
---------------
Bb:ion    	->	['C', 'D♭', 'D', 'E♭', 'E', 'F', 'G♭', 'G', 'A♭', 'A', 'B♭', 'B']	->	['B♭', 'C', 'D', 'E♭', 'F', 'G', 'A']
Bb:dor    	->	['C', 'D♭', 'D', 'E♭', 'E', 'F', 'G♭', 'G', 'A♭', 'A', 'B♭', 'B']	->	['A♭', 'B♭', 'C', 'D♭', 'E♭', 'F', 'G']
Bb:phr    	->	['C', 'D♭', 'D', 'E♭', 'E', 'F', 'G♭', 'G', 'A♭', 'A', 'B♭', 'C♭']	->	['G♭', 'A♭', 'B♭', 'C♭', 'D♭', 'E♭', 'F']
Bb:lyd    	->	['C', 'D♭', 'D', 'E♭', 'E', 'F', 'G♭', 'G', 'A♭', 'A', 'B♭', 'B']	->	['F', 'G', 'A', 'B♭', 'C', 'D', 'E']
Bb:mix    	->	['C', 'D♭', 'D', 'E♭', 'E', 'F', 'G♭', 'G', 'A♭', 'A', 'B♭', 'B']	->	['E♭', 'F', 'G', 'A♭', 'B♭', 'C', 'D']
Bb:aeol   	->	['C', 'D♭', 'D', 'E♭', 'E', 'F', 'G♭', 'G', 'A♭', 'A', 'B♭', 'B']	->	['D♭', 'E♭', 'F', 'G♭', 'A♭', 'B♭', 'C']
Bb:loc    	->	['C', 'D♭', 'D', 'E♭', 'F♭', 'F', 'G♭', 'G', 'A♭', 'A', 'B♭', 'C♭']	->	['C♭', 'D♭', 'E♭', 'F♭', 'G♭', 'A♭', 'B♭']
---------------
B:ion     	->	['C', 'C♯', 'D', 'D♯', 'E', 'F', 'F♯', 'G', 'G♯', 'A', 'A♯', 'B']	->	['B', 'C♯', 'D♯', 'E', 'F♯', 'G♯', 'A♯']
B:dor     	->	['C', 'C♯', 'D', 'D♯', 'E', 'F', 'F♯', 'G', 'G♯', 'A', 'A♯', 'B']	->	['A', 'B', 'C♯', 'D', 'E', 'F♯', 'G♯']
B:phr     	->	['C', 'C♯', 'D', 'D♯', 'E', 'F', 'F♯', 'G', 'G♯', 'A', 'A♯', 'B']	->	['G', 'A', 'B', 'C', 'D', 'E', 'F♯']
B:lyd     	->	['C', 'C♯', 'D', 'D♯', 'E', 'E♯', 'F♯', 'G', 'G♯', 'A', 'A♯', 'B']	->	['F♯', 'G♯', 'A♯', 'B', 'C♯', 'D♯', 'E♯']
B:mix     	->	['C', 'C♯', 'D', 'D♯', 'E', 'F', 'F♯', 'G', 'G♯', 'A', 'A♯', 'B']	->	['E', 'F♯', 'G♯', 'A', 'B', 'C♯', 'D♯']
B:aeol    	->	['C', 'C♯', 'D', 'D♯', 'E', 'F', 'F♯', 'G', 'G♯', 'A', 'A♯', 'B']	->	['D', 'E', 'F♯', 'G', 'A', 'B', 'C♯']
B:loc     	->	['C', 'C♯', 'D', 'D♯', 'E', 'F', 'F♯', 'G', 'G♯', 'A', 'A♯', 'B']	->	['C', 'D', 'E', 'F', 'G', 'A', 'B']
---------------
B#:ion    	->	['B♯', 'C♯', 'C𝄪', 'D♯', 'D𝄪', 'E♯', 'F♯', 'F𝄪', 'G♯', 'G𝄪', 'A♯', 'A𝄪']	->	['B♯', 'C𝄪', 'D𝄪', 'E♯', 'F𝄪', 'G𝄪', 'A𝄪']
B#:dor    	->	['B♯', 'C♯', 'C𝄪', 'D♯', 'E', 'E♯', 'F♯', 'F𝄪', 'G♯', 'G𝄪', 'A♯', 'B']	->	['A♯', 'B♯', 'C𝄪', 'D♯', 'E♯', 'F𝄪', 'G𝄪']
B#:phr    	->	['B♯', 'C♯', 'D', 'D♯', 'E', 'E♯', 'F♯', 'F𝄪', 'G♯', 'A', 'A♯', 'B']	->	['G♯', 'A♯', 'B♯', 'C♯', 'D♯', 'E♯', 'F𝄪']
B#:lyd    	->	['B♯', 'C♯', 'C𝄪', 'D♯', 'D𝄪', 'E♯', 'E𝄪', 'F𝄪', 'G♯', 'G𝄪', 'A♯', 'A𝄪']	->	['F𝄪', 'G𝄪', 'A𝄪', 'B♯', 'C𝄪', 'D𝄪', 'E𝄪']
B#:mix    	->	['B♯', 'C♯', 'C𝄪', 'D♯', 'D𝄪', 'E♯', 'F♯', 'F𝄪', 'G♯', 'G𝄪', 'A♯', 'B']	->	['E♯', 'F𝄪', 'G𝄪', 'A♯', 'B♯', 'C𝄪', 'D𝄪']
B#:aeol   	->	['B♯', 'C♯', 'C𝄪', 'D♯', 'E', 'E♯', 'F♯', 'F𝄪', 'G♯', 'A', 'A♯', 'B']	->	['D♯', 'E♯', 'F𝄪', 'G♯', 'A♯', 'B♯', 'C𝄪']
B#:loc    	->	['B♯', 'C♯', 'D', 'D♯', 'E', 'E♯', 'F♯', 'G', 'G♯', 'A', 'A♯', 'B']	->	['C♯', 'D♯', 'E♯', 'F♯', 'G♯', 'A♯', 'B♯']
```
---------------
</details>

erabinov commented:
Ok, I'll try to get the delinting done this afternoon (CA time).

For the modes, if I understand correctly, you have issues only with the `key_to_degrees` function, since when you input, say `C:phr`, you just get the corresponding major scale, say the A-flat major scale?

This seems like it could be quickly fixed by adding a dictionary that maps modes to the number of positions we have to rotate the array by. Happy to try to hack it all together.

erabinov commented:
Ok, hopefully this most recent commit will pass the linting! It's very exciting that the modes are working as expected.

bmcfee commented:
> Ok, I'll try to get the delinting done this afternoon (CA time).
> 
> For the modes, if I understand correctly, you have issues only with the `key_to_degrees` function, since when you input, say `C:phr`, you just get the corresponding major scale, say the A-flat major scale?
> 
> This seems like it could be quickly fixed by adding a dictionary that maps modes to the number of positions we have to rotate the array by. Happy to try to hack it all together.

Yup, that's exactly it.  Easy enough to fix as a post-hoc rotation, and doing so looks as below.  The part that I/we need to think about is where this belongs in the code, and how it will look from the API perspective.  Happy to handle that in a separate PR though.


```python
for k in 'CDEFGAB':
    for acc in ['bb', 'b', '', '#', '##']:
        root = k + acc
        for offset, mode in enumerate(['ion', 'dor', 'phr', 'lyd', 'mix', 'aeol', 'loc']):
            query = f"{root}:{mode}"
            equiv = mode_to_key(query)

            degs = np.roll(librosa.key_to_degrees(equiv), -offset)
            notes = librosa.key_to_notes(equiv)
            
            names = [notes[_] for _ in degs]
            #print(f"{query:10s}\t→\t{notes}\t→\t{names}", end='')
            print(f"{query:10s}\t→\t{names}", end='')
            if len(set([n[0] for n in names])) != 7:
                print('\t->\tFAIL')
            else:
                print()
        print('-' * 15)
```

<details>
<summary> Revised outputs, with double-accidentals even! </summary>

```
Cbb:ion   	→	['C𝄫', 'D𝄫', 'E𝄫', 'F𝄫', 'G𝄫', 'A𝄫', 'B𝄫']
Cbb:dor   	→	['C𝄫', 'D𝄫', 'E♭𝄫', 'F𝄫', 'G𝄫', 'A𝄫', 'B♭𝄫']
Cbb:phr   	→	['C𝄫', 'D♭𝄫', 'E♭𝄫', 'F𝄫', 'G𝄫', 'A♭𝄫', 'B♭𝄫']
Cbb:lyd   	→	['C𝄫', 'D𝄫', 'E𝄫', 'F♭', 'G𝄫', 'A𝄫', 'B𝄫']
Cbb:mix   	→	['C𝄫', 'D𝄫', 'E𝄫', 'F𝄫', 'G𝄫', 'A𝄫', 'B♭𝄫']
Cbb:aeol  	→	['C𝄫', 'D𝄫', 'E♭𝄫', 'F𝄫', 'G𝄫', 'A♭𝄫', 'B♭𝄫']
Cbb:loc   	→	['C𝄫', 'D♭𝄫', 'E♭𝄫', 'F𝄫', 'G♭𝄫', 'A♭𝄫', 'B♭𝄫']
---------------
Cb:ion    	→	['C♭', 'D♭', 'E♭', 'F♭', 'G♭', 'A♭', 'B♭']
Cb:dor    	→	['C♭', 'D♭', 'E𝄫', 'F♭', 'G♭', 'A♭', 'B𝄫']
Cb:phr    	→	['C♭', 'D𝄫', 'E𝄫', 'F♭', 'G♭', 'A𝄫', 'B𝄫']
Cb:lyd    	→	['C♭', 'D♭', 'E♭', 'F', 'G♭', 'A♭', 'B♭']
Cb:mix    	→	['C♭', 'D♭', 'E♭', 'F♭', 'G♭', 'A♭', 'B𝄫']
Cb:aeol   	→	['C♭', 'D♭', 'E𝄫', 'F♭', 'G♭', 'A𝄫', 'B𝄫']
Cb:loc    	→	['C♭', 'D𝄫', 'E𝄫', 'F♭', 'G𝄫', 'A𝄫', 'B𝄫']
---------------
C:ion     	→	['C', 'D', 'E', 'F', 'G', 'A', 'B']
C:dor     	→	['C', 'D', 'E♭', 'F', 'G', 'A', 'B♭']
C:phr     	→	['C', 'D♭', 'E♭', 'F', 'G', 'A♭', 'B♭']
C:lyd     	→	['C', 'D', 'E', 'F♯', 'G', 'A', 'B']
C:mix     	→	['C', 'D', 'E', 'F', 'G', 'A', 'B♭']
C:aeol    	→	['C', 'D', 'E♭', 'F', 'G', 'A♭', 'B♭']
C:loc     	→	['C', 'D♭', 'E♭', 'F', 'G♭', 'A♭', 'B♭']
---------------
C#:ion    	→	['C♯', 'D♯', 'E♯', 'F♯', 'G♯', 'A♯', 'B♯']
C#:dor    	→	['C♯', 'D♯', 'E', 'F♯', 'G♯', 'A♯', 'B']
C#:phr    	→	['C♯', 'D', 'E', 'F♯', 'G♯', 'A', 'B']
C#:lyd    	→	['C♯', 'D♯', 'E♯', 'F𝄪', 'G♯', 'A♯', 'B♯']
C#:mix    	→	['C♯', 'D♯', 'E♯', 'F♯', 'G♯', 'A♯', 'B']
C#:aeol   	→	['C♯', 'D♯', 'E', 'F♯', 'G♯', 'A', 'B']
C#:loc    	→	['C♯', 'D', 'E', 'F♯', 'G', 'A', 'B']
---------------
C##:ion   	→	['C𝄪', 'D𝄪', 'E𝄪', 'F𝄪', 'G𝄪', 'A𝄪', 'B𝄪']
C##:dor   	→	['C𝄪', 'D𝄪', 'E♯', 'F𝄪', 'G𝄪', 'A𝄪', 'B♯']
C##:phr   	→	['C𝄪', 'D♯', 'E♯', 'F𝄪', 'G𝄪', 'A♯', 'B♯']
C##:lyd   	→	['C𝄪', 'D𝄪', 'E𝄪', 'F♯𝄪', 'G𝄪', 'A𝄪', 'B𝄪']
C##:mix   	→	['C𝄪', 'D𝄪', 'E𝄪', 'F𝄪', 'G𝄪', 'A𝄪', 'B♯']
C##:aeol  	→	['C𝄪', 'D𝄪', 'E♯', 'F𝄪', 'G𝄪', 'A♯', 'B♯']
C##:loc   	→	['C𝄪', 'D♯', 'E♯', 'F𝄪', 'G♯', 'A♯', 'B♯']
---------------
Dbb:ion   	→	['D𝄫', 'E𝄫', 'F♭', 'G𝄫', 'A𝄫', 'B𝄫', 'C♭']
Dbb:dor   	→	['D𝄫', 'E𝄫', 'F𝄫', 'G𝄫', 'A𝄫', 'B𝄫', 'C𝄫']
Dbb:phr   	→	['D𝄫', 'E♭𝄫', 'F𝄫', 'G𝄫', 'A𝄫', 'B♭𝄫', 'C𝄫']
Dbb:lyd   	→	['D𝄫', 'E𝄫', 'F♭', 'G♭', 'A𝄫', 'B𝄫', 'C♭']
Dbb:mix   	→	['D𝄫', 'E𝄫', 'F♭', 'G𝄫', 'A𝄫', 'B𝄫', 'C𝄫']
Dbb:aeol  	→	['D𝄫', 'E𝄫', 'F𝄫', 'G𝄫', 'A𝄫', 'B♭𝄫', 'C𝄫']
Dbb:loc   	→	['D𝄫', 'E♭𝄫', 'F𝄫', 'G𝄫', 'A♭𝄫', 'B♭𝄫', 'C𝄫']
---------------
Db:ion    	→	['D♭', 'E♭', 'F', 'G♭', 'A♭', 'B♭', 'C']
Db:dor    	→	['D♭', 'E♭', 'F♭', 'G♭', 'A♭', 'B♭', 'C♭']
Db:phr    	→	['D♭', 'E𝄫', 'F♭', 'G♭', 'A♭', 'B𝄫', 'C♭']
Db:lyd    	→	['D♭', 'E♭', 'F', 'G', 'A♭', 'B♭', 'C']
Db:mix    	→	['D♭', 'E♭', 'F', 'G♭', 'A♭', 'B♭', 'C♭']
Db:aeol   	→	['D♭', 'E♭', 'F♭', 'G♭', 'A♭', 'B𝄫', 'C♭']
Db:loc    	→	['D♭', 'E𝄫', 'F♭', 'G♭', 'A𝄫', 'B𝄫', 'C♭']
---------------
D:ion     	→	['D', 'E', 'F♯', 'G', 'A', 'B', 'C♯']
D:dor     	→	['D', 'E', 'F', 'G', 'A', 'B', 'C']
D:phr     	→	['D', 'E♭', 'F', 'G', 'A', 'B♭', 'C']
D:lyd     	→	['D', 'E', 'F♯', 'G♯', 'A', 'B', 'C♯']
D:mix     	→	['D', 'E', 'F♯', 'G', 'A', 'B', 'C']
D:aeol    	→	['D', 'E', 'F', 'G', 'A', 'B♭', 'C']
D:loc     	→	['D', 'E♭', 'F', 'G', 'A♭', 'B♭', 'C']
---------------
D#:ion    	→	['D♯', 'E♯', 'F𝄪', 'G♯', 'A♯', 'B♯', 'C𝄪']
D#:dor    	→	['D♯', 'E♯', 'F♯', 'G♯', 'A♯', 'B♯', 'C♯']
D#:phr    	→	['D♯', 'E', 'F♯', 'G♯', 'A♯', 'B', 'C♯']
D#:lyd    	→	['D♯', 'E♯', 'F𝄪', 'G𝄪', 'A♯', 'B♯', 'C𝄪']
D#:mix    	→	['D♯', 'E♯', 'F𝄪', 'G♯', 'A♯', 'B♯', 'C♯']
D#:aeol   	→	['D♯', 'E♯', 'F♯', 'G♯', 'A♯', 'B', 'C♯']
D#:loc    	→	['D♯', 'E', 'F♯', 'G♯', 'A', 'B', 'C♯']
---------------
D##:ion   	→	['D𝄪', 'E𝄪', 'F♯𝄪', 'G𝄪', 'A𝄪', 'B𝄪', 'C♯𝄪']
D##:dor   	→	['D𝄪', 'E𝄪', 'F𝄪', 'G𝄪', 'A𝄪', 'B𝄪', 'C𝄪']
D##:phr   	→	['D𝄪', 'E♯', 'F𝄪', 'G𝄪', 'A𝄪', 'B♯', 'C𝄪']
D##:lyd   	→	['D𝄪', 'E𝄪', 'F♯𝄪', 'G♯𝄪', 'A𝄪', 'B𝄪', 'C♯𝄪']
D##:mix   	→	['D𝄪', 'E𝄪', 'F♯𝄪', 'G𝄪', 'A𝄪', 'B𝄪', 'C𝄪']
D##:aeol  	→	['D𝄪', 'E𝄪', 'F𝄪', 'G𝄪', 'A𝄪', 'B♯', 'C𝄪']
D##:loc   	→	['D𝄪', 'E♯', 'F𝄪', 'G𝄪', 'A♯', 'B♯', 'C𝄪']
---------------
Ebb:ion   	→	['E𝄫', 'F♭', 'G♭', 'A𝄫', 'B𝄫', 'C♭', 'D♭']
Ebb:dor   	→	['E𝄫', 'F♭', 'G𝄫', 'A𝄫', 'B𝄫', 'C♭', 'D𝄫']
Ebb:phr   	→	['E𝄫', 'F𝄫', 'G𝄫', 'A𝄫', 'B𝄫', 'C𝄫', 'D𝄫']
Ebb:lyd   	→	['E𝄫', 'F♭', 'G♭', 'A♭', 'B𝄫', 'C♭', 'D♭']
Ebb:mix   	→	['E𝄫', 'F♭', 'G♭', 'A𝄫', 'B𝄫', 'C♭', 'D𝄫']
Ebb:aeol  	→	['E𝄫', 'F♭', 'G𝄫', 'A𝄫', 'B𝄫', 'C𝄫', 'D𝄫']
Ebb:loc   	→	['E𝄫', 'F𝄫', 'G𝄫', 'A𝄫', 'B♭𝄫', 'C𝄫', 'D𝄫']
---------------
Eb:ion    	→	['E♭', 'F', 'G', 'A♭', 'B♭', 'C', 'D']
Eb:dor    	→	['E♭', 'F', 'G♭', 'A♭', 'B♭', 'C', 'D♭']
Eb:phr    	→	['E♭', 'F♭', 'G♭', 'A♭', 'B♭', 'C♭', 'D♭']
Eb:lyd    	→	['E♭', 'F', 'G', 'A', 'B♭', 'C', 'D']
Eb:mix    	→	['E♭', 'F', 'G', 'A♭', 'B♭', 'C', 'D♭']
Eb:aeol   	→	['E♭', 'F', 'G♭', 'A♭', 'B♭', 'C♭', 'D♭']
Eb:loc    	→	['E♭', 'F♭', 'G♭', 'A♭', 'B𝄫', 'C♭', 'D♭']
---------------
E:ion     	→	['E', 'F♯', 'G♯', 'A', 'B', 'C♯', 'D♯']
E:dor     	→	['E', 'F♯', 'G', 'A', 'B', 'C♯', 'D']
E:phr     	→	['E', 'F', 'G', 'A', 'B', 'C', 'D']
E:lyd     	→	['E', 'F♯', 'G♯', 'A♯', 'B', 'C♯', 'D♯']
E:mix     	→	['E', 'F♯', 'G♯', 'A', 'B', 'C♯', 'D']
E:aeol    	→	['E', 'F♯', 'G', 'A', 'B', 'C', 'D']
E:loc     	→	['E', 'F', 'G', 'A', 'B♭', 'C', 'D']
---------------
E#:ion    	→	['E♯', 'F𝄪', 'G𝄪', 'A♯', 'B♯', 'C𝄪', 'D𝄪']
E#:dor    	→	['E♯', 'F𝄪', 'G♯', 'A♯', 'B♯', 'C𝄪', 'D♯']
E#:phr    	→	['E♯', 'F♯', 'G♯', 'A♯', 'B♯', 'C♯', 'D♯']
E#:lyd    	→	['E♯', 'F𝄪', 'G𝄪', 'A𝄪', 'B♯', 'C𝄪', 'D𝄪']
E#:mix    	→	['E♯', 'F𝄪', 'G𝄪', 'A♯', 'B♯', 'C𝄪', 'D♯']
E#:aeol   	→	['E♯', 'F𝄪', 'G♯', 'A♯', 'B♯', 'C♯', 'D♯']
E#:loc    	→	['E♯', 'F♯', 'G♯', 'A♯', 'B', 'C♯', 'D♯']
---------------
E##:ion   	→	['E𝄪', 'F♯𝄪', 'G♯𝄪', 'A𝄪', 'B𝄪', 'C♯𝄪', 'D♯𝄪']
E##:dor   	→	['E𝄪', 'F♯𝄪', 'G𝄪', 'A𝄪', 'B𝄪', 'C♯𝄪', 'D𝄪']
E##:phr   	→	['E𝄪', 'F𝄪', 'G𝄪', 'A𝄪', 'B𝄪', 'C𝄪', 'D𝄪']
E##:lyd   	→	['E𝄪', 'F♯𝄪', 'G♯𝄪', 'A♯𝄪', 'B𝄪', 'C♯𝄪', 'D♯𝄪']
E##:mix   	→	['E𝄪', 'F♯𝄪', 'G♯𝄪', 'A𝄪', 'B𝄪', 'C♯𝄪', 'D𝄪']
E##:aeol  	→	['E𝄪', 'F♯𝄪', 'G𝄪', 'A𝄪', 'B𝄪', 'C𝄪', 'D𝄪']
E##:loc   	→	['E𝄪', 'F𝄪', 'G𝄪', 'A𝄪', 'B♯', 'C𝄪', 'D𝄪']
---------------
Fbb:ion   	→	['F𝄫', 'G𝄫', 'A𝄫', 'B♭𝄫', 'C𝄫', 'D𝄫', 'E𝄫']
Fbb:dor   	→	['F𝄫', 'G𝄫', 'A♭𝄫', 'B♭𝄫', 'C𝄫', 'D𝄫', 'E♭𝄫']
Fbb:phr   	→	['F𝄫', 'G♭𝄫', 'A♭𝄫', 'B♭𝄫', 'C𝄫', 'D♭𝄫', 'E♭𝄫']
Fbb:lyd   	→	['F𝄫', 'G𝄫', 'A𝄫', 'B𝄫', 'C𝄫', 'D𝄫', 'E𝄫']
Fbb:mix   	→	['F𝄫', 'G𝄫', 'A𝄫', 'B♭𝄫', 'C𝄫', 'D𝄫', 'E♭𝄫']
Fbb:aeol  	→	['F𝄫', 'G𝄫', 'A♭𝄫', 'B♭𝄫', 'C𝄫', 'D♭𝄫', 'E♭𝄫']
Fbb:loc   	→	['F𝄫', 'G♭𝄫', 'A♭𝄫', 'B♭𝄫', 'C♭𝄫', 'D♭𝄫', 'E♭𝄫']
---------------
Fb:ion    	→	['F♭', 'G♭', 'A♭', 'B𝄫', 'C♭', 'D♭', 'E♭']
Fb:dor    	→	['F♭', 'G♭', 'A𝄫', 'B𝄫', 'C♭', 'D♭', 'E𝄫']
Fb:phr    	→	['F♭', 'G𝄫', 'A𝄫', 'B𝄫', 'C♭', 'D𝄫', 'E𝄫']
Fb:lyd    	→	['F♭', 'G♭', 'A♭', 'B♭', 'C♭', 'D♭', 'E♭']
Fb:mix    	→	['F♭', 'G♭', 'A♭', 'B𝄫', 'C♭', 'D♭', 'E𝄫']
Fb:aeol   	→	['F♭', 'G♭', 'A𝄫', 'B𝄫', 'C♭', 'D𝄫', 'E𝄫']
Fb:loc    	→	['F♭', 'G𝄫', 'A𝄫', 'B𝄫', 'C𝄫', 'D𝄫', 'E𝄫']
---------------
F:ion     	→	['F', 'G', 'A', 'B♭', 'C', 'D', 'E']
F:dor     	→	['F', 'G', 'A♭', 'B♭', 'C', 'D', 'E♭']
F:phr     	→	['F', 'G♭', 'A♭', 'B♭', 'C', 'D♭', 'E♭']
F:lyd     	→	['F', 'G', 'A', 'B', 'C', 'D', 'E']
F:mix     	→	['F', 'G', 'A', 'B♭', 'C', 'D', 'E♭']
F:aeol    	→	['F', 'G', 'A♭', 'B♭', 'C', 'D♭', 'E♭']
F:loc     	→	['F', 'G♭', 'A♭', 'B♭', 'C♭', 'D♭', 'E♭']
---------------
F#:ion    	→	['F♯', 'G♯', 'A♯', 'B', 'C♯', 'D♯', 'E♯']
F#:dor    	→	['F♯', 'G♯', 'A', 'B', 'C♯', 'D♯', 'E']
F#:phr    	→	['F♯', 'G', 'A', 'B', 'C♯', 'D', 'E']
F#:lyd    	→	['F♯', 'G♯', 'A♯', 'B♯', 'C♯', 'D♯', 'E♯']
F#:mix    	→	['F♯', 'G♯', 'A♯', 'B', 'C♯', 'D♯', 'E']
F#:aeol   	→	['F♯', 'G♯', 'A', 'B', 'C♯', 'D', 'E']
F#:loc    	→	['F♯', 'G', 'A', 'B', 'C', 'D', 'E']
---------------
F##:ion   	→	['F𝄪', 'G𝄪', 'A𝄪', 'B♯', 'C𝄪', 'D𝄪', 'E𝄪']
F##:dor   	→	['F𝄪', 'G𝄪', 'A♯', 'B♯', 'C𝄪', 'D𝄪', 'E♯']
F##:phr   	→	['F𝄪', 'G♯', 'A♯', 'B♯', 'C𝄪', 'D♯', 'E♯']
F##:lyd   	→	['F𝄪', 'G𝄪', 'A𝄪', 'B𝄪', 'C𝄪', 'D𝄪', 'E𝄪']
F##:mix   	→	['F𝄪', 'G𝄪', 'A𝄪', 'B♯', 'C𝄪', 'D𝄪', 'E♯']
F##:aeol  	→	['F𝄪', 'G𝄪', 'A♯', 'B♯', 'C𝄪', 'D♯', 'E♯']
F##:loc   	→	['F𝄪', 'G♯', 'A♯', 'B♯', 'C♯', 'D♯', 'E♯']
---------------
Gbb:ion   	→	['G𝄫', 'A𝄫', 'B𝄫', 'C𝄫', 'D𝄫', 'E𝄫', 'F♭']
Gbb:dor   	→	['G𝄫', 'A𝄫', 'B♭𝄫', 'C𝄫', 'D𝄫', 'E𝄫', 'F𝄫']
Gbb:phr   	→	['G𝄫', 'A♭𝄫', 'B♭𝄫', 'C𝄫', 'D𝄫', 'E♭𝄫', 'F𝄫']
Gbb:lyd   	→	['G𝄫', 'A𝄫', 'B𝄫', 'C♭', 'D𝄫', 'E𝄫', 'F♭']
Gbb:mix   	→	['G𝄫', 'A𝄫', 'B𝄫', 'C𝄫', 'D𝄫', 'E𝄫', 'F𝄫']
Gbb:aeol  	→	['G𝄫', 'A𝄫', 'B♭𝄫', 'C𝄫', 'D𝄫', 'E♭𝄫', 'F𝄫']
Gbb:loc   	→	['G𝄫', 'A♭𝄫', 'B♭𝄫', 'C𝄫', 'D♭𝄫', 'E♭𝄫', 'F𝄫']
---------------
Gb:ion    	→	['G♭', 'A♭', 'B♭', 'C♭', 'D♭', 'E♭', 'F']
Gb:dor    	→	['G♭', 'A♭', 'B𝄫', 'C♭', 'D♭', 'E♭', 'F♭']
Gb:phr    	→	['G♭', 'A𝄫', 'B𝄫', 'C♭', 'D♭', 'E𝄫', 'F♭']
Gb:lyd    	→	['G♭', 'A♭', 'B♭', 'C', 'D♭', 'E♭', 'F']
Gb:mix    	→	['G♭', 'A♭', 'B♭', 'C♭', 'D♭', 'E♭', 'F♭']
Gb:aeol   	→	['G♭', 'A♭', 'B𝄫', 'C♭', 'D♭', 'E𝄫', 'F♭']
Gb:loc    	→	['G♭', 'A𝄫', 'B𝄫', 'C♭', 'D𝄫', 'E𝄫', 'F♭']
---------------
G:ion     	→	['G', 'A', 'B', 'C', 'D', 'E', 'F♯']
G:dor     	→	['G', 'A', 'B♭', 'C', 'D', 'E', 'F']
G:phr     	→	['G', 'A♭', 'B♭', 'C', 'D', 'E♭', 'F']
G:lyd     	→	['G', 'A', 'B', 'C♯', 'D', 'E', 'F♯']
G:mix     	→	['G', 'A', 'B', 'C', 'D', 'E', 'F']
G:aeol    	→	['G', 'A', 'B♭', 'C', 'D', 'E♭', 'F']
G:loc     	→	['G', 'A♭', 'B♭', 'C', 'D♭', 'E♭', 'F']
---------------
G#:ion    	→	['G♯', 'A♯', 'B♯', 'C♯', 'D♯', 'E♯', 'F𝄪']
G#:dor    	→	['G♯', 'A♯', 'B', 'C♯', 'D♯', 'E♯', 'F♯']
G#:phr    	→	['G♯', 'A', 'B', 'C♯', 'D♯', 'E', 'F♯']
G#:lyd    	→	['G♯', 'A♯', 'B♯', 'C𝄪', 'D♯', 'E♯', 'F𝄪']
G#:mix    	→	['G♯', 'A♯', 'B♯', 'C♯', 'D♯', 'E♯', 'F♯']
G#:aeol   	→	['G♯', 'A♯', 'B', 'C♯', 'D♯', 'E', 'F♯']
G#:loc    	→	['G♯', 'A', 'B', 'C♯', 'D', 'E', 'F♯']
---------------
G##:ion   	→	['G𝄪', 'A𝄪', 'B𝄪', 'C𝄪', 'D𝄪', 'E𝄪', 'F♯𝄪']
G##:dor   	→	['G𝄪', 'A𝄪', 'B♯', 'C𝄪', 'D𝄪', 'E𝄪', 'F𝄪']
G##:phr   	→	['G𝄪', 'A♯', 'B♯', 'C𝄪', 'D𝄪', 'E♯', 'F𝄪']
G##:lyd   	→	['G𝄪', 'A𝄪', 'B𝄪', 'C♯𝄪', 'D𝄪', 'E𝄪', 'F♯𝄪']
G##:mix   	→	['G𝄪', 'A𝄪', 'B𝄪', 'C𝄪', 'D𝄪', 'E𝄪', 'F𝄪']
G##:aeol  	→	['G𝄪', 'A𝄪', 'B♯', 'C𝄪', 'D𝄪', 'E♯', 'F𝄪']
G##:loc   	→	['G𝄪', 'A♯', 'B♯', 'C𝄪', 'D♯', 'E♯', 'F𝄪']
---------------
Abb:ion   	→	['A𝄫', 'B𝄫', 'C♭', 'D𝄫', 'E𝄫', 'F♭', 'G♭']
Abb:dor   	→	['A𝄫', 'B𝄫', 'C𝄫', 'D𝄫', 'E𝄫', 'F♭', 'G𝄫']
Abb:phr   	→	['A𝄫', 'B♭𝄫', 'C𝄫', 'D𝄫', 'E𝄫', 'F𝄫', 'G𝄫']
Abb:lyd   	→	['A𝄫', 'B𝄫', 'C♭', 'D♭', 'E𝄫', 'F♭', 'G♭']
Abb:mix   	→	['A𝄫', 'B𝄫', 'C♭', 'D𝄫', 'E𝄫', 'F♭', 'G𝄫']
Abb:aeol  	→	['A𝄫', 'B𝄫', 'C𝄫', 'D𝄫', 'E𝄫', 'F𝄫', 'G𝄫']
Abb:loc   	→	['A𝄫', 'B♭𝄫', 'C𝄫', 'D𝄫', 'E♭𝄫', 'F𝄫', 'G𝄫']
---------------
Ab:ion    	→	['A♭', 'B♭', 'C', 'D♭', 'E♭', 'F', 'G']
Ab:dor    	→	['A♭', 'B♭', 'C♭', 'D♭', 'E♭', 'F', 'G♭']
Ab:phr    	→	['A♭', 'B𝄫', 'C♭', 'D♭', 'E♭', 'F♭', 'G♭']
Ab:lyd    	→	['A♭', 'B♭', 'C', 'D', 'E♭', 'F', 'G']
Ab:mix    	→	['A♭', 'B♭', 'C', 'D♭', 'E♭', 'F', 'G♭']
Ab:aeol   	→	['A♭', 'B♭', 'C♭', 'D♭', 'E♭', 'F♭', 'G♭']
Ab:loc    	→	['A♭', 'B𝄫', 'C♭', 'D♭', 'E𝄫', 'F♭', 'G♭']
---------------
A:ion     	→	['A', 'B', 'C♯', 'D', 'E', 'F♯', 'G♯']
A:dor     	→	['A', 'B', 'C', 'D', 'E', 'F♯', 'G']
A:phr     	→	['A', 'B♭', 'C', 'D', 'E', 'F', 'G']
A:lyd     	→	['A', 'B', 'C♯', 'D♯', 'E', 'F♯', 'G♯']
A:mix     	→	['A', 'B', 'C♯', 'D', 'E', 'F♯', 'G']
A:aeol    	→	['A', 'B', 'C', 'D', 'E', 'F', 'G']
A:loc     	→	['A', 'B♭', 'C', 'D', 'E♭', 'F', 'G']
---------------
A#:ion    	→	['A♯', 'B♯', 'C𝄪', 'D♯', 'E♯', 'F𝄪', 'G𝄪']
A#:dor    	→	['A♯', 'B♯', 'C♯', 'D♯', 'E♯', 'F𝄪', 'G♯']
A#:phr    	→	['A♯', 'B', 'C♯', 'D♯', 'E♯', 'F♯', 'G♯']
A#:lyd    	→	['A♯', 'B♯', 'C𝄪', 'D𝄪', 'E♯', 'F𝄪', 'G𝄪']
A#:mix    	→	['A♯', 'B♯', 'C𝄪', 'D♯', 'E♯', 'F𝄪', 'G♯']
A#:aeol   	→	['A♯', 'B♯', 'C♯', 'D♯', 'E♯', 'F♯', 'G♯']
A#:loc    	→	['A♯', 'B', 'C♯', 'D♯', 'E', 'F♯', 'G♯']
---------------
A##:ion   	→	['A𝄪', 'B𝄪', 'C♯𝄪', 'D𝄪', 'E𝄪', 'F♯𝄪', 'G♯𝄪']
A##:dor   	→	['A𝄪', 'B𝄪', 'C𝄪', 'D𝄪', 'E𝄪', 'F♯𝄪', 'G𝄪']
A##:phr   	→	['A𝄪', 'B♯', 'C𝄪', 'D𝄪', 'E𝄪', 'F𝄪', 'G𝄪']
A##:lyd   	→	['A𝄪', 'B𝄪', 'C♯𝄪', 'D♯𝄪', 'E𝄪', 'F♯𝄪', 'G♯𝄪']
A##:mix   	→	['A𝄪', 'B𝄪', 'C♯𝄪', 'D𝄪', 'E𝄪', 'F♯𝄪', 'G𝄪']
A##:aeol  	→	['A𝄪', 'B𝄪', 'C𝄪', 'D𝄪', 'E𝄪', 'F𝄪', 'G𝄪']
A##:loc   	→	['A𝄪', 'B♯', 'C𝄪', 'D𝄪', 'E♯', 'F𝄪', 'G𝄪']
---------------
Bbb:ion   	→	['B𝄫', 'C♭', 'D♭', 'E𝄫', 'F♭', 'G♭', 'A♭']
Bbb:dor   	→	['B𝄫', 'C♭', 'D𝄫', 'E𝄫', 'F♭', 'G♭', 'A𝄫']
Bbb:phr   	→	['B𝄫', 'C𝄫', 'D𝄫', 'E𝄫', 'F♭', 'G𝄫', 'A𝄫']
Bbb:lyd   	→	['B𝄫', 'C♭', 'D♭', 'E♭', 'F♭', 'G♭', 'A♭']
Bbb:mix   	→	['B𝄫', 'C♭', 'D♭', 'E𝄫', 'F♭', 'G♭', 'A𝄫']
Bbb:aeol  	→	['B𝄫', 'C♭', 'D𝄫', 'E𝄫', 'F♭', 'G𝄫', 'A𝄫']
Bbb:loc   	→	['B𝄫', 'C𝄫', 'D𝄫', 'E𝄫', 'F𝄫', 'G𝄫', 'A𝄫']
---------------
Bb:ion    	→	['B♭', 'C', 'D', 'E♭', 'F', 'G', 'A']
Bb:dor    	→	['B♭', 'C', 'D♭', 'E♭', 'F', 'G', 'A♭']
Bb:phr    	→	['B♭', 'C♭', 'D♭', 'E♭', 'F', 'G♭', 'A♭']
Bb:lyd    	→	['B♭', 'C', 'D', 'E', 'F', 'G', 'A']
Bb:mix    	→	['B♭', 'C', 'D', 'E♭', 'F', 'G', 'A♭']
Bb:aeol   	→	['B♭', 'C', 'D♭', 'E♭', 'F', 'G♭', 'A♭']
Bb:loc    	→	['B♭', 'C♭', 'D♭', 'E♭', 'F♭', 'G♭', 'A♭']
---------------
B:ion     	→	['B', 'C♯', 'D♯', 'E', 'F♯', 'G♯', 'A♯']
B:dor     	→	['B', 'C♯', 'D', 'E', 'F♯', 'G♯', 'A']
B:phr     	→	['B', 'C', 'D', 'E', 'F♯', 'G', 'A']
B:lyd     	→	['B', 'C♯', 'D♯', 'E♯', 'F♯', 'G♯', 'A♯']
B:mix     	→	['B', 'C♯', 'D♯', 'E', 'F♯', 'G♯', 'A']
B:aeol    	→	['B', 'C♯', 'D', 'E', 'F♯', 'G', 'A']
B:loc     	→	['B', 'C', 'D', 'E', 'F', 'G', 'A']
---------------
B#:ion    	→	['B♯', 'C𝄪', 'D𝄪', 'E♯', 'F𝄪', 'G𝄪', 'A𝄪']
B#:dor    	→	['B♯', 'C𝄪', 'D♯', 'E♯', 'F𝄪', 'G𝄪', 'A♯']
B#:phr    	→	['B♯', 'C♯', 'D♯', 'E♯', 'F𝄪', 'G♯', 'A♯']
B#:lyd    	→	['B♯', 'C𝄪', 'D𝄪', 'E𝄪', 'F𝄪', 'G𝄪', 'A𝄪']
B#:mix    	→	['B♯', 'C𝄪', 'D𝄪', 'E♯', 'F𝄪', 'G𝄪', 'A♯']
B#:aeol   	→	['B♯', 'C𝄪', 'D♯', 'E♯', 'F𝄪', 'G♯', 'A♯']
B#:loc    	→	['B♯', 'C♯', 'D♯', 'E♯', 'F♯', 'G♯', 'A♯']
---------------
B##:ion   	→	['B𝄪', 'C♯𝄪', 'D♯𝄪', 'E𝄪', 'F♯𝄪', 'G♯𝄪', 'A♯𝄪']
B##:dor   	→	['B𝄪', 'C♯𝄪', 'D𝄪', 'E𝄪', 'F♯𝄪', 'G♯𝄪', 'A𝄪']
B##:phr   	→	['B𝄪', 'C𝄪', 'D𝄪', 'E𝄪', 'F♯𝄪', 'G𝄪', 'A𝄪']
B##:lyd   	→	['B𝄪', 'C♯𝄪', 'D♯𝄪', 'E♯𝄪', 'F♯𝄪', 'G♯𝄪', 'A♯𝄪']
B##:mix   	→	['B𝄪', 'C♯𝄪', 'D♯𝄪', 'E𝄪', 'F♯𝄪', 'G♯𝄪', 'A𝄪']
B##:aeol  	→	['B𝄪', 'C♯𝄪', 'D𝄪', 'E𝄪', 'F♯𝄪', 'G𝄪', 'A𝄪']
B##:loc   	→	['B𝄪', 'C𝄪', 'D𝄪', 'E𝄪', 'F𝄪', 'G𝄪', 'A𝄪']
---------------

```

</details>

bmcfee commented on librosa/core/notation.py:
```diff
@@ -471,6 +473,93 @@ def list_thaat() -> List[str]:
+
+    """
+    if not isinstance(key,str):
+        return np.array([__simplify_note(n+additional_acc, unicode=unicode) for n in key])
```
Can you add a test case that covers this path, ie simplifying an array of keys?

erabinov replied:
Sounds good. I'll get right on it!

By the way, is there a way I can do the lint and codecov checking locally so that I don't have to keep triggering the tests before I know whether they will pass?

bmcfee replied:
Yeah!  If you check the CONTRIBUTING guidelines, it should give you everything you need to set up a local test environment.  Then it should just be a matter of running the same commands that are in our github actions:
https://github.com/librosa/librosa/blob/09e4a622456bbad314e2e6d5c75879f56d728b9b/.github/workflows/lint_python.yml#L45-L71
and
https://github.com/librosa/librosa/blob/09e4a622456bbad314e2e6d5c75879f56d728b9b/.github/workflows/ci.yml#L106

It won't upload the coverage report to codecov, but you'll at least be able to see the coverage report locally.

bmcfee reviewed (changes requested):
I think this is pretty much there - thanks so much for the persistence!

I have one last request here, to make sure we maintain full code path coverage.  Once that's in, I think we're all set.

bmcfee commented:
Thanks again @erabinov - this is great!

I'll leave the original issue open until the second half (mode support) gets rolled in, but this goes a huge step of the way toward unblocking that.

erabinov commented:
Thanks for your patience in working with a first-time contributor!!

Let me know if you could use some help with implementing support for modes.

bmcfee commented:
> Thanks for your patience in working with a first-time contributor!!
> 
> Let me know if you could use some help with implementing support for modes.

That would be great, if you're up for it!

The code I posted above in thread should be a pretty self-explanatory starting point, and as far as i can tell, it all works perfectly thanks to this pr.  The part that will need some thought is how it will integrate with the existing functions, what the api will look like, and how we might need to refactor existing functions.

erabinov commented:
> That would be great, if you're up for it!
> 
> The code I posted above in thread should be a pretty self-explanatory starting point, and as far as i can tell, it all works perfectly thanks to this pr. The part that will need some thought is how it will integrate with the existing functions, what the api will look like, and how we might need to refactor existing functions.

Ok, I'll try to get a sense of what goes wrong after a "simple copy-paste" this weekend or early next week. It's a good excuse to get my local environment in fit enough shape to run the delinting, etc on my own.
